daemon/src/setup_taker.rs

```diff
@@ -0,0 +1,252 @@
+        );
+
+        let (sender, receiver) = mpsc::unbounded::<SetupMsg>();
+        // store the writing end to forward messages from the taker to
+        // the spawned contract setup task
+        self.setup_msg_sender = Some(sender);
+
+        let contract_future = setup_contract::new(
+            self.maker
+                .sink()
+                .with(move |msg| future::ok(wire::TakerToMaker::Protocol { order_id, msg })),
+            receiver,
+            (self.oracle_pk, self.announcement.clone()),
+            SetupParams::new(
+                cfd.margin()?,
+                cfd.counterparty_margin()?,
+                cfd.order.price,
+                cfd.quantity_usd,
+                cfd.order.leverage,
+                cfd.refund_timelock_in_blocks(),
+            ),
+            self.build_party_params.clone_channel(),
+            self.sign.clone_channel(),
+            Role::Taker,
+            self.n_payouts,
+        );
+
+        let this = ctx.address().expect("self to be alive");
+        spawn_fallible::<_, anyhow::Error>(async move {
+            let _ = match contract_future.await {
+                Ok(dlc) => this.send(SetupSucceeded { order_id, dlc }).await?,
+                Err(error) => this.send(SetupFailed { order_id, error }).await?,
+            };
+
+            Ok(())
+        });
+
+        Ok(())
+    }
+
+    fn handle(&mut self, _: Rejected, ctx: &mut xtra::Context<Self>) -> Result<()> {
+        self.on_completed
+            .send(Completed::Rejected {
+                order_id: self.order.id,
+            })
+            .await?;
+
+        ctx.stop();
+
+        Ok(())
+    }
+
+    fn handle(&mut self, msg: wire::SetupMsg, _ctx: &mut xtra::Context<Self>) -> Result<()> {
+        let mut sender = self
+            .setup_msg_sender
+            .clone()
+            .context("Cannot forward message to contract setup task")?;
+        sender.send(msg).await?;
+
+        Ok(())
+    }
+
+    fn handle(&mut self, msg: SetupSucceeded, ctx: &mut xtra::Context<Self>) -> Result<()> {
+        self.on_completed
+            .send(Completed::NewContract {
+                order_id: msg.order_id,
+                dlc: msg.dlc,
+            })
+            .await?;
+
+        ctx.stop();
+
+        Ok(())
+    }
+
+    fn handle(&mut self, msg: SetupFailed, ctx: &mut xtra::Context<Self>) -> Result<()> {
+        self.on_completed
+            .send(Completed::Failed {
+                order_id: msg.order_id,
+                error: msg.error,
+            })
+            .await?;
+
+        ctx.stop();
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    async fn started(&mut self, ctx: &mut xtra::Context<Self>) {
+        let order_id = self.order.id;
+        self.maker
+            .send(wire::TakerToMaker::TakeOrder {
+                order_id,
+                quantity: self.quantity,
+            })
+            .await
+            .expect("connection actor to be up");
```

Comment:
I considered something like what you describe at first, but decided against it because I thought that taking an `Address` to the `connection::Actor` would force us to change the `taker_cfd::Actor` in ways that would break the actor tests. Admittedly, I didn't give it a go, so maybe I was wrong about that. I'll give it a go now.